float32 xParam
float32 yParam
float32 zParam
float32 dParam
